### PR TITLE
Training model

### PR DESCRIPTION
**Pull request by 51N84D**
Model is trainable, with comet logging and saving/restoring checkpoints

### CODE REVIEW

File: models/base_model.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+            "cuda" if (torch.cuda.is_available() and opt.model.use_gpu) else "cpu"
+        )
+        self.isTrain = opt.model.is_train
+        self.loss_names = []
+        self.model_names = []
+        self.comet_exp = opt.comet_exp
+        self.save_dir = opt.train.output_dir + "/checkpoints"
+
+    def set_input(self, input):
+        pass
```
Comment (vict0rsch) [inline models/base_model.py]: It's a detail but if the meaning of these `pass` statements is to say that this must be implemented by classes which inherit from `BaseModel` then you should have `raise NotImplementedError("implement set_input")` instead

File: models/base_model.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    # intermediate steps for backprop
+    def test(self):
+        with torch.no_grad():
+            self.forward()
+
+    def optimize_parameters(self):
+        pass
+
+    # update learning rate (called once every epoch)
+    def update_learning_rate(self):
```
Comment (vict0rsch) [inline models/base_model.py]: This is syntactically correct but you're lucky the docstring counts ; this should `raise` something or at least `pass`

File: models/base_model.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    # save models to the disk
+    def save_networks(self, epoch):
+        print("save models")  # TODO: save checkpoints
+        for name in self.model_names:
+            if isinstance(name, str):
+                save_filename = "%s_net_%s.pth" % (epoch, name)
+                save_path = os.path.join(self.save_dir, save_filename)
+                net = getattr(self, "net" + name)
+
+                if self.use_gpu and torch.cuda.is_available():
```
Comment (vict0rsch) [inline models/base_model.py]: You may have inverted the conditions here

File: models/mask_generator.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        self.netD = networks.define_D(opt).to(self.device)
+        self.comet_exp = opt.comet.exp
+        self.store_image = opt.val.store_image
+
+        if self.isTrain:
+            # define loss functions
+            self.criterionGAN = networks.GANLoss().to(self.device)
+            # initialize optimizers
+
+            self.optimizer_G = torch.optim.Adam(
```
Comment (vict0rsch) [inline models/mask_generator.py]: you could have an option to use ExtraAdam (see Omnigan)

File: models/networks.py [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+
+    print("initialize network with %s" % init_type)
+    net.apply(init_func)
+
+
+def l2normalize(v, eps=1e-12):
+    return v / (v.norm() + eps)
+
+
+def get_norm_layer(norm_type="instance"):
```
Comment (vict0rsch) [inline models/networks.py]: see omnigan issue assigned to @51N84D in order to include spectral norm

File: models/networks.py [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+    elif norm_type == "instance":
+        norm_layer = functools.partial(nn.InstanceNorm2d, affine=False, track_running_stats=False)
+    elif norm_type == "none":
+        norm_layer = None
+    else:
+        raise NotImplementedError("normalization layer [%s] is not found" % norm_type)
+    return norm_layer
+
+
+class GANLoss(nn.Module):
```
Comment (vict0rsch) [inline models/networks.py]: introduce label flipping? (see omnigan)

File: models/networks.py [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+        self.encoder = nn.Sequential(*self.encoder)
+
+        # --------------------DECODER----------------------------
+        self.decoder = []
+        for i in range(n_downsample):
+            self.decoder += [
+                ConvTranspose2dBlock(dim, int(dim / 2), 2, 2, 0, norm=dec_norm, activation=activ)
+            ]
+            dim = int(dim / 2)
+        self.decoder += [Conv2dBlock(dim, output_dim, 3, 1, 1, activation=output_activ)]
```
Comment (vict0rsch) [inline models/networks.py]: shouldn't `output_activ` always be `tanh`?

File: tests/test_dataloaders.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
+import os
```
Comment (vict0rsch) [inline tests/test_dataloaders.py]: yeeey tests :)

File: train.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+    val_opt = set_mode("test", opt)
+    val_loader = get_loader(val_opt)
+    dataset_size = len(loader)
+    print("#training images = %d" % dataset_size)
+
+    comet_exp = Experiment(workspace=opt.comet.workspace, project_name=opt.comet.project_name)
+    if comet_exp is not None:
+        comet_exp.log_asset(file_data=str(root / opt_file), file_name=root / opt_file)
+        comet_exp.log_parameters(opt)
```
Comment (vict0rsch) [inline train.py]: so comet handles multi-level (nested) dicts? cool

File: utils.py [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+    """
+
+    for mode in ["train", "val"]:
+        for domain in opts.data.files[mode]:
+            opts.data.files[mode] = str(Path(opts.data.files.base) / opts.data.files[mode])
+
+    return opts
+
+
+def set_mode(mode, opts):
```
Comment (vict0rsch) [inline utils.py]: remember to have docstrings even for such a simple function

**Comment by vict0rsch** [comment]
(and delete branch when no longer working on it to keep repo clean)
